File: packages/indexer/cognis_indexer/enricher/secrets.py
"""Secret detector for the cognis enricher pipeline.

Implements two complementary detection strategies:

1. **Known-shape regex patterns** — AWS access keys, GitHub PATs, Slack tokens,
   Google API keys, OpenAI keys, JWTs, PEM private-key headers, DSNs with
   embedded credentials, and password/secret assignment patterns.

2. **Shannon-entropy threshold** — string literals that exceed 4.5 bits/char
   AND are ≥ 16 characters long are flagged as high-entropy secrets.

The public interface is :class:`SecretDetector` with a single
:meth:`~SecretDetector.redact` method:

    >>> sd = SecretDetector()
    >>> redacted, types = sd.redact("key = 'AKIAIOSFODNN7EXAMPLE'")
    >>> assert "[REDACTED:aws-access-key]" in redacted
    >>> assert "aws-access-key" in types

Secrets are replaced in-place with ``[REDACTED:<type>]``.  The original string
is **never** persisted.

Design reference: design.md *Indexer Pipeline → Enricher → Secret detection*.
"""

from __future__ import annotations

import re
from collections import Counter
from math import log2

# ---------------------------------------------------------------------------
# Known-shape patterns
# ---------------------------------------------------------------------------
# Each entry: (compiled_pattern, label_used_in_[REDACTED:label])
# Patterns are ordered from most specific to least specific to avoid
# "almost matches" swallowing a more precise pattern.

_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    # AWS access key id (AKIA prefix + 16 uppercase alphanumeric chars)
    (re.compile(r"AKIA[0-9A-Z]{16}"), "aws-access-key"),
    # GitHub personal access token (ghp_ prefix)
    (re.compile(r"ghp_[A-Za-z0-9]{36}"), "github-pat"),
    # Slack tokens (xoxb/xoxa/xoxp/xoxr/xoxs)
    (re.compile(r"xox[baprs]-[A-Za-z0-9-]{10,}"), "slack-token"),
    # Google API key (AIza prefix)
    (re.compile(r"AIza[0-9A-Za-z\-_]{35}"), "google-api-key"),
    # OpenAI secret key / project key (sk- or sk-proj- prefix)
    (re.compile(r"sk-(?:proj-)?[A-Za-z0-9]{20,}"), "openai-key"),
    # JWT — three base64url segments separated by dots
    (
        re.compile(
            r"eyJ[A-Za-z0-9_=-]+"
            r"\.[A-Za-z0-9_=-]+"
            r"\.?[A-Za-z0-9_.+/=-]*"
        ),
        "jwt",
    ),
    # PEM private key header (BEGIN ... PRIVATE KEY)
    (
        re.compile(r"-----BEGIN [A-Z ]+PRIVATE KEY-----"),
        "pem-private-key-header",
    ),
    # DSN / URL with embedded credentials: scheme://[user]:pass@host
    # Supports empty username (redis://:pass@host) and non-empty (pg://user:pass@host)
    (
        re.compile(
            r"[a-zA-Z][a-zA-Z0-9+.\-]*://"  # scheme
            r"[^/\s:@]*"  # username (may be empty)
            r":"  # colon
            r"[^/\s:@]+"  # password (non-empty, no slashes/spaces/@)
            r"@"  # at-sign
            r"[^\s'\"#]+"  # host + path (no quotes/hash)
        ),
        "dsn-with-credentials",
    ),
]

# Password / secret assignment: password = "value", passwd = 'value', etc.
# We capture group 1 = keyword, group 2 = value so we can reconstruct the
# assignment with the value replaced.
_PASSWORD_RE = re.compile(
    r"""(?ix)
    \b
    (password|passwd|pwd|secret|api[_\-]?key|token)  # keyword
    \s*[:=]\s*                                         # separator
    ['"]                                               # opening quote
    ([^'"\n]{4,})                                      # value (4+ chars)
    ['"]                                               # closing quote
    """,
)

# ---------------------------------------------------------------------------
# Entropy helpers
# ---------------------------------------------------------------------------


def _shannon_entropy(value: str) -> float:
    """Return Shannon entropy in bits per character for *value*.

    Uses the standard formula ``-sum(p * log2(p))`` where *p* is the
    probability of each unique character.

    Returns ``0.0`` for empty strings.
    """
    if not value:
        return 0.0
    counts = Counter(value)
    total = float(len(value))
    return -sum((n / total) * log2(n / total) for n in counts.values())


def _is_high_entropy(value: str, threshold: float = 4.5) -> bool:
    """Return True when *value* looks like a secret based on entropy alone.

    Both conditions must hold:
    - ``len(value) >= 16`` (short strings are exempt).
    - Shannon entropy >= *threshold* (default 4.5 bits/char).
    """
    return len(value) >= 16 and _shannon_entropy(value) >= threshold


# Pattern to find quoted string literals (used for entropy scan).
_QUOTED_STRING_RE = re.compile(
    r"""(?x)
    (?:
        '([^'\n]{16,})'       # single-quoted, 16+ chars
    |
        "([^"\n]{16,})"       # double-quoted, 16+ chars
    )
    """,
)

# ---------------------------------------------------------------------------
# SecretDetector
# ---------------------------------------------------------------------------


class SecretDetector:
    """Detect and redact secret-shaped strings.

    This class is stateless; a single instance is safe to share across threads.

    Example::

        sd = SecretDetector()
        clean, types = sd.redact(text)
        if types:
            # at least one secret was found and redacted
            ...
    """

    def redact(self, text: str) -> tuple[str, list[str]]:
        """Replace all secret-shaped substrings in *text* with ``[REDACTED:<type>]``.

        Applies pattern matching first, then entropy-based detection on any
        remaining quoted string literals.

        Args:
            text: Input string (body_excerpt, signature, docstring, etc.).

        Returns:
            A 2-tuple ``(redacted_text, types_found)`` where:

            - *redacted_text* is *text* with every secret replaced by its
              ``[REDACTED:<type>]`` placeholder.  If no secrets are found,
              *redacted_text* is the original *text* unchanged.
            - *types_found* is a list of unique redaction-type labels (e.g.
              ``["aws-access-key", "jwt"]``).  Order reflects first occurrence.
              Empty list when no secrets found.
        """
        if not text:
            return text, []

        out = text
        found_types: list[str] = []
        seen_types: set[str] = set()

        def _record(label: str) -> None:
            if label not in seen_types:
                seen_types.add(label)
                found_types.append(label)

        # 1. Apply known-shape regex patterns
        for pattern, label in _PATTERNS:
            new_out, count = pattern.subn(f"[REDACTED:{label}]", out)
            if count > 0:
                _record(label)
                out = new_out

        # 2. Apply password/secret assignment pattern
        def _replace_password(m: re.Match[str]) -> str:
            _record("password-assignment")
            return f'{m.group(1)}="[REDACTED:password-assignment]"'

        out = _PASSWORD_RE.sub(_replace_password, out)

        # 3. Entropy-based scan on any quoted strings that survived step 1+2
        def _replace_high_entropy(m: re.Match[str]) -> str:
            # group(1) = single-quoted value, group(2) = double-quoted value
            value = m.group(1) or m.group(2) or ""
            if _is_high_entropy(value):
                _record("high-entropy")
                quote = "'" if m.group(1) is not None else '"'
                return f"{quote}[REDACTED:high-entropy]{quote}"
            return m.group(0)

        out = _QUOTED_STRING_RE.sub(_replace_high_entropy, out)

        return out, found_types

    # ------------------------------------------------------------------
    # Convenience helpers
    # ------------------------------------------------------------------

    def is_secret(self, value: str) -> bool:
        """Return True when *value* matches any known secret pattern.

        This is a fast check (no redaction); useful for validation tests.
        """
        if not value:
            return False
        return any(p.search(value) for p, _ in _PATTERNS) or bool(_PASSWORD_RE.search(value))

    def shannon_entropy(self, value: str) -> float:
        """Expose :func:`_shannon_entropy` as a public method for testing."""
        return _shannon_entropy(value)

    def is_high_entropy(self, value: str, threshold: float = 4.5) -> bool:
        """Expose :func:`_is_high_entropy` as a public method for testing."""
        return _is_high_entropy(value, threshold)

File: crates/cognis-indexer/src/enricher/secrets.rs
//! Secret detector for the enricher stage (Task 8.2, Requirement 9.3).
//!
//! Rust mirror of `cognis_indexer.enricher.secrets.SecretDetector`. Two
//! complementary strategies, applied in order:
//!
//! 1. **Known-shape regex patterns** — AWS access keys, GitHub PATs, Slack
//!    tokens, Google API keys, OpenAI keys, JWTs, PEM private-key headers, DSNs
//!    with embedded credentials, then password/secret assignment.
//! 2. **Shannon-entropy threshold** — any remaining quoted string literal that
//!    is ≥ 16 chars and ≥ 4.5 bits/char of entropy is flagged.
//!
//! Each match is replaced in place with `[REDACTED:<type>]`; the original
//! string is **never** returned. The enricher runs this **before** the Writer
//! persists anything (Requirement 9.3).

use std::collections::HashMap;
use std::sync::OnceLock;

use regex::Regex;

/// High-entropy threshold in bits per character.
const ENTROPY_THRESHOLD: f64 = 4.5;
/// Minimum length for the entropy check to apply.
const ENTROPY_MIN_LEN: usize = 16;

/// Known-shape `(regex, label)` patterns, most specific first (mirror
/// `secrets._PATTERNS`).
fn patterns() -> &'static [(Regex, &'static str)] {
    static PATS: OnceLock<Vec<(Regex, &'static str)>> = OnceLock::new();
    PATS.get_or_init(|| {
        vec![
            (Regex::new(r"AKIA[0-9A-Z]{16}").unwrap(), "aws-access-key"),
            (Regex::new(r"ghp_[A-Za-z0-9]{36}").unwrap(), "github-pat"),
            (
                Regex::new(r"xox[baprs]-[A-Za-z0-9-]{10,}").unwrap(),
                "slack-token",
            ),
            (
                Regex::new(r"AIza[0-9A-Za-z\-_]{35}").unwrap(),
                "google-api-key",
            ),
            (
                Regex::new(r"sk-(?:proj-)?[A-Za-z0-9]{20,}").unwrap(),
                "openai-key",
            ),
            (
                Regex::new(r"eyJ[A-Za-z0-9_=-]+\.[A-Za-z0-9_=-]+\.?[A-Za-z0-9_.+/=-]*").unwrap(),
                "jwt",
            ),
            (
                Regex::new(r"-----BEGIN [A-Z ]+PRIVATE KEY-----").unwrap(),
                "pem-private-key-header",
            ),
            (
                Regex::new(r##"[a-zA-Z][a-zA-Z0-9+.\-]*://[^/\s:@]*:[^/\s:@]+@[^\s'"#]+"##)
                    .unwrap(),
                "dsn-with-credentials",
            ),
        ]
    })
}

/// `password = "..."` / `token: '...'` assignment (mirror `_PASSWORD_RE`).
fn password_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r#"(?i)\b(password|passwd|pwd|secret|api[_\-]?key|token)\s*[:=]\s*["']([^"'\n]{4,})["']"#)
            .unwrap()
    })
}

/// Quoted string literal of 16+ chars (single or double) for the entropy scan.
fn quoted_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r#"'([^'\n]{16,})'|"([^"\n]{16,})""#).unwrap())
}

/// Detect and redact secret-shaped strings. Stateless; safe to share.
#[derive(Debug, Default)]
pub struct SecretDetector;

impl SecretDetector {
    /// Replace every secret-shaped substring in `text` with `[REDACTED:<type>]`.
    ///
    /// Returns `(redacted_text, types_found)` where `types_found` lists unique
    /// redaction labels in first-occurrence order. When nothing matches the
    /// original text is returned unchanged with an empty list.
    pub fn redact(&self, text: &str) -> (String, Vec<String>) {
        if text.is_empty() {
            return (text.to_string(), Vec::new());
        }

        let mut out = text.to_string();
        let mut found: Vec<String> = Vec::new();
        let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
        let mut record = |label: &str, found: &mut Vec<String>| {
            if seen.insert(label.to_string()) {
                found.push(label.to_string());
            }
        };

        // 1. Known-shape regex patterns (sequential, each over the running text).
        for (pattern, label) in patterns() {
            if pattern.is_match(&out) {
                record(label, &mut found);
                out = pattern
                    .replace_all(&out, format!("[REDACTED:{label}]").as_str())
                    .into_owned();
            }
        }

        // 2. Password / secret assignment — keep the keyword, redact the value.
        if password_re().is_match(&out) {
            record("password-assignment", &mut found);
            out = password_re()
                .replace_all(&out, |caps: &regex::Captures| {
                    format!("{}=\"[REDACTED:password-assignment]\"", &caps[1])
                })
                .into_owned();
        }

        // 3. Entropy scan on any quoted strings that survived steps 1+2.
        let mut high_entropy_seen = false;
        out = quoted_re()
            .replace_all(&out, |caps: &regex::Captures| {
                let (value, quote) = match (caps.get(1), caps.get(2)) {
                    (Some(m), _) => (m.as_str(), '\''),
                    (_, Some(m)) => (m.as_str(), '"'),
                    _ => return caps[0].to_string(),
                };
                if is_high_entropy(value) {
                    high_entropy_seen = true;
                    format!("{quote}[REDACTED:high-entropy]{quote}")
                } else {
                    caps[0].to_string()
                }
            })
            .into_owned();
        if high_entropy_seen {
            record("high-entropy", &mut found);
        }

        (out, found)
    }
}

/// Shannon entropy (bits per character) of `value`; `0.0` for empty input.
pub fn shannon_entropy(value: &str) -> f64 {
    if value.is_empty() {
        return 0.0;
    }
    let mut counts: HashMap<char, usize> = HashMap::new();
    let mut total = 0.0f64;
    for ch in value.chars() {
        *counts.entry(ch).or_insert(0) += 1;
        total += 1.0;
    }
    -counts
        .values()
        .map(|&n| {
            let p = n as f64 / total;
            p * p.log2()
        })
        .sum::<f64>()
}

/// `true` when `value` looks like a secret on entropy alone (≥ 16 chars AND
/// entropy ≥ 4.5 bits/char).
pub fn is_high_entropy(value: &str) -> bool {
    value.chars().count() >= ENTROPY_MIN_LEN && shannon_entropy(value) >= ENTROPY_THRESHOLD
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn redacts_aws_access_key() {
        let (out, types) = SecretDetector.redact("key = 'AKIAIOSFODNN7EXAMPLE'");
        assert!(out.contains("[REDACTED:aws-access-key]"));
        assert!(types.contains(&"aws-access-key".to_string()));
    }

    #[test]
    fn redacts_password_assignment_keeps_keyword() {
        let (out, types) = SecretDetector.redact(r#"password = "hunter2secret""#);
        assert!(out.contains("password=\"[REDACTED:password-assignment]\""));
        assert_eq!(types, vec!["password-assignment".to_string()]);
    }

    #[test]
    fn redacts_dsn_with_credentials() {
        let (out, types) = SecretDetector.redact("url = 'postgres://user:s3cr3t@db:5432/app'");
        assert!(out.contains("[REDACTED:dsn-with-credentials]"));
        assert!(types.contains(&"dsn-with-credentials".to_string()));
    }

    #[test]
    fn high_entropy_quoted_string_flagged() {
        // 24 distinct chars → entropy log2(24) ≈ 4.58 ≥ 4.5 threshold.
        let secret = "aB3dE6gH9jK2mN5pQ8sT1vW4";
        let (out, types) = SecretDetector.redact(&format!("nonce = '{secret}'"));
        assert!(types.contains(&"high-entropy".to_string()));
        assert!(out.contains("[REDACTED:high-entropy]"));
    }

    #[test]
    fn clean_text_unchanged() {
        let text = "def add(a, b):\n    return a + b";
        let (out, types) = SecretDetector.redact(text);
        assert_eq!(out, text);
        assert!(types.is_empty());
    }

    #[test]
    fn empty_entropy_is_zero() {
        assert_eq!(shannon_entropy(""), 0.0);
        assert!(!is_high_entropy("short"));
    }
}
